### ISSUE TITLE
Can not configure samba printer

### ISSUE THREAD

**Issue by gjask** [fedora-silverblue/issue-tracker#532]
**Describe the bug**
**tl; dr:** CUPS web console misses option to add new printer via SAMBA on Silverblue

At my workplace we have this network printer we connect using SAMBA. For past seven years I always configure this printer using web console natively provided by CUPS (http://localhost:631) because in comparison to gnome-settings it always works like a charm. Gnome-settings have changed many times its printers configuration dialogues in that time yet there is always some problem preventing to configure this particular printer (could not connect samba, could not load custom printer driver, could not handle samba auth, etc.) .

Today I tried to configure this printer in same way on my new laptop for which I decided to switch to Silverblue and I could not because CUPS web console misses option to add samba printer. I spun up Workstation in Boxes to check if that is also an issue there, but it seems to be Silverblue related.

**To Reproduce**
Please describe the steps needed to reproduce the bug:
1. Open http://localhost:631 in browser
2. Navigate to `Administration` and authenticate with your system account
3. Click on `Add Printer` in `Printers` paragraph
4. Try to chose  `Windows Printer via SAMBA`

**Expected behavior**
Option `Windows Printer via SAMBA` is offered and working

**Screenshots**
**Silverblue**:
![Snímek obrazovky z 2024-01-30 17-59-43](https://github.com/fedora-silverblue/issue-tracker/assets/3306648/493de539-840e-4962-86c1-7ed769268a2f)

**Workstation**:
![Snímek obrazovky z 2024-01-30 17-59-04](https://github.com/fedora-silverblue/issue-tracker/assets/3306648/5d3b8168-94d9-4709-ae8d-e0c7eafae4d6)

**OS version:**
```
State: idle
BootedDeployment:
● fedora:fedora/39/x86_64/silverblue
                  Version: 39.20240129.0 (2024-01-29T05:35:53Z)
               BaseCommit: bdd0afb451c1768bf872aa30c7d6ee5767f72fd75a75275a54a75f6b30804b61
             GPGSignature: Valid signature by E8F23996F23218640CB44CBE75CF5AC418B8E74C
      RemovedBasePackages: libavfilter-free libavformat-free libpostproc-free libswresample-free libavutil-free libswscale-free libavcodec-free 6.0.1-2.fc39
                           mesa-va-drivers 23.3.3-1.fc39 nano-default-editor 7.2-4.fc39
          LayeredPackages: adw-gtk3-theme ffmpeg gnome-tweaks htop langpacks-cs
                           mesa-va-drivers-freeworld rpmfusion-free-release
                           rpmfusion-nonfree-release tilix vim-default-editor wl-clipboard
```

**Comment by travier** [comment]
This is because we don't have `samba` in Siverblue. This is related to https://github.com/fedora-silverblue/issue-tracker/issues/320.

I think we should re-consider including it.

**Comment by gjask** [comment]
> This is because we don't have `samba` in Siverblue.

That's actually first thing I considered but I also did quick check and it seemed not be the issue.

```bash
# Silverblue
jask@silverbook:~$ rpm -qa | grep samba
samba-common-4.19.4-3.fc39.noarch
samba-client-libs-4.19.4-3.fc39.x86_64
samba-common-libs-4.19.4-3.fc39.x86_64
samba-client-4.19.4-3.fc39.x86_64

# Workstation
jask@fedora:~$ rpm -qa | grep samba
samba-common-4.19.4-3.fc39.noarch
samba-common-libs-4.19.4-3.fc39.x86_64
samba-client-libs-4.19.4-3.fc39.x86_64
samba-client-4.19.4-3.fc39.x86_64
```

Of course it doesn't mean some other required package that is part of Workstation is not part of Silverblue. But I thought there should be functional/package parity between Workstation and Silverblue, shouldn't it?

**Comment by gjask** [comment]
After poking around for some time I was able to come up with cause to this issue. According to [CUPS docs] there should be `/usr/lib/cups/backend/smb` symlink pointing to `smbspool`. I've confirmed it enables configuring SAMBA printer by creating this symlink in usroverlay.

This symlink is present in Workstation installation but is missing in Silverblue. I tracked the missing file to this line in samba package: https://src.fedoraproject.org/rpms/cups/blob/rawhide/f/cups.spec#_570

I am not sure why this doesn't get applied in Silverblue, because `samba-client` seems to be present.

[CUPS docs]: https://www.cups.org/blog/2005-06-17-printing-to-windows-servers-with-samba.html

**Comment by gjask** [comment]
This could be related to https://github.com/coreos/rpm-ostree/issues/1180

Or it can be unrelated as none of the greps cgwalters used to find affected packages wouldn't match this script.

**Comment by travier** [comment]
Nice investigation, this looks like the issue indeed. Not sure if we have a good path forward. Firsts step would be to confirm if we have SAMBA or not in Silverblue by default right now.

**Comment by gjask** [comment]
> Nice investigation, this looks like the issue indeed. Not sure if we have a good path forward. 

I think there should be two options. Either implement `%triggerin` in rpm-ostree. Or refactor spec file for `cups` to not need it.

Second option can actually be easier and faster to implement. However it doesn't prevent similar issues in future.

> Firsts step would be to confirm if we have SAMBA or not in Silverblue by default right now.

I am not sure how to properly confirm it. To me it seems that same samba packages are part of Silverblue as are installed in Workstation.

**Comment by gjask** [comment]
> Firsts step would be to confirm if we have SAMBA or not in Silverblue by default right now.

It seems we do: https://pagure.io/workstation-ostree-config/blob/f39/f/fedora-common-ostree-pkgs.yaml#_151

**Comment by LorbusChris** [comment]
Couldn't (or shouldn't really) this symlink be shipped as a tmpfiles config snippet? The trigger could then just be `systemd-tmpfiles --create`. Might have other ramifications, but might also be worth opening a PR for on the spec to discuss with the RPM maintainers.

**Comment by travier** [comment]
Could you clarify how we could fix that with tmpfiles?

I've made a temporary workaround in https://pagure.io/workstation-ostree-config/pull-request/456. @gjask can  you confirm that the paths are correct? Thanks

**Comment by gjask** [comment]
> I've made a temporary workaround in https://pagure.io/workstation-ostree-config/pull-request/456. @gjask can you confirm that the paths are correct? Thanks

This looks like it should work for most use-cases. But it would likely break for people who would want to layer and use `samba-krb5-printing`. I think correct symlink is the one the original trigger tries to create. It should be as follows

```bash
ln -snf /usr/libexec/samba/cups_backend_smb /usr/lib/cups/backend/smb
```

Apart from that it seems it should work. Thank you for the idea how to hotfix this. :)

**Comment by travier** [comment]
Thanks for the review. I've updated the PR.

**Comment by travier** [comment]
I've merged the PR for Rawhide and I'm backporting this in F39 in https://pagure.io/workstation-ostree-config/pull-request/458

I'll merge it soon so that's in the next update. Can you test it tomorrow and report back? Thanks!

**Comment by gjask** [comment]
@travier It seems it is working to the point I am able to add the printer as expected. I still am not able print, but for now it seems to be unrelated problem as same happens on current Workstation. I'll look into it more.

**Comment by travier** [comment]
Thanks for the update! I'll keep this one open to track this issue.

**Comment by travier** [comment]
Not sure I understand. You can not write to `/usr` on Silverblue.

**Comment by LorbusChris** [comment]
ah right, nevermind this isn't going to work then.

**Comment by gjask** [comment]
> Is this still an issue in Fedora 41?

You mean if there is still need for this workaround? https://pagure.io/workstation-ostree-config/blob/f41/f/common.yaml#_168

There still is `%triggerin` in CUPS [package](https://src.fedoraproject.org/rpms/cups/blob/f41/f/cups.spec#_554) and upstream issue https://github.com/coreos/rpm-ostree/issues/1180 seems still unresolved. So I guess yes?

**Comment by travier** [comment]
That workaround will indeed stay there for a while. My question was more about whether you could get your printer working?

**Comment by gjask** [comment]
Oh, yes. The workaround indeed worked. The issues I faced later were unrelated to Silverblue and were resolved elsewhere so I didn't report on them here.

**Comment by travier** [comment]
Thanks, let's close this then.
